akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -3005,25 +3012,67 @@ trait FlowOps[+Out, +Mat] {
    * '''Cancels when''' downstream cancels
    */
   def concat[U >: Out, Mat2](that: Graph[SourceShape[U], Mat2]): Repr[U] =
-    via(concatGraph(that))
+    internalConcat(that, detached = true)
 
   protected def concatGraph[U >: Out, Mat2](
-      that: Graph[SourceShape[U], Mat2]): Graph[FlowShape[Out @uncheckedVariance, U], Mat2] =
+      that: Graph[SourceShape[U], Mat2],
+      detached: Boolean): Graph[FlowShape[Out @uncheckedVariance, U], Mat2] =
     GraphDSL.create(that) { implicit b => r =>
-      val merge = b.add(Concat[U]())
+      val merge = b.add(Concat[U](2, detached))
       r ~> merge.in(1)
       FlowShape(merge.in(0), merge.out)
```

Comment:
Is the graph builder api efficient enough (with not too many allocations)? Would it be worth having something more specialized for constructing this? Probably not worth it until we see it in profiling, but just a thought.

Reply:
For the concat-single we got away without it at least, I think that is a very common case.

Reply:
GraphDSL definitely turns up in profiles but as Johan says, if we can avoid building any graphs it's even better.